### PR TITLE
Add util function to get SHAP values via CV

### PR DESCRIPTION
**Pull request by rchaves33**
# Motivation

SHAP values provide an excellent way to estimate model feature importance. The SHAP values tell us how much a feature contributes to (impacts) the model prediction.

When we evaluate model performance, we typically want to know the performance on different data sets:

* train
* validation (val)
* test

Similarly, when we calculate SHAP values, we want to know the impact of the features when the model is applied to different data sets.

With the current `shap` API, it is straightforward to calculate SHAP values for train, val, and test sets.

However, most modeling problems involve limited data, so designating both a val and test set is unacceptable, because too little data is left to train the model. Enter cross validation, of course, where there is no hold-out val set, but the validation occurs via CV on the train set.

With the current `shap` API, there is no support for calculating SHAP values via CV (e.g. by integration with `sklearn`).

So we have three options:

* Calculate SHAP values on train. But this is as flawed as evaluating model performance on train alone. You will get the importance of features of a model trained and evaluated on the same data, i.e. it is not a fair evaluation.
* Calculate SHAP values on test. This seems like a better choice, but if we are calculating SHAP values as a way of selecting model features, it seems a dangerous choice. Ideally, modeling decisions should never be based on the hold-out test set -- that's why we have val sets and CV after all.
* Calculate SHAP values on train via CV. Sounds ideal! But how?
This PR provides a helper function to calculate SHAP values via CV.

Possible contribution to `probatus.utils.shap_helpers` if `probatus` maintainers see value and agree.

`ShapRFECV` is still the better option overall, but for some users a recursive approach may not be feasible due to computational cost. With `get_shap_cv_value`, user gets the benefit of CV in a non-recursive approach and can then select the `n` top features in more confidence.

### CODE REVIEW

**Comment by rchaves33** [comment]
Shoutout to @L-Ramos: we discovered your similar work while discussing this code. Glad to see others are thinking along the same lines!

https://lucasramos-34338.medium.com/visualizing-variable-importance-using-shap-and-cross-validation-bd5075e9063a

**Comment by L-Ramos** [comment]
@rchaves33 This will be a very valuable addition to the library and make things a lot easier. Very nice work!

File: probatus/utils/shap_helpers.py [hunk 1]
```diff
@@ -200,3 +201,66 @@ def calculate_shap_importance(shap_values, columns, output_columns_suffix=""):
...
+
+    Returns:
+        shap_values_cv
+    """
+    if isinstance(cv, int):
+        cv = StratifiedKFold(n_splits=cv, shuffle=True, random_state=0)
+
+    shap_values_cv = []
+
+    for train_index, test_index in cv.split(X_train, y_train):
```
Comment (Matgrb) [inline probatus/utils/shap_helpers.py:239]: One quick thing to test if it works with groups like here https://ing-bank.github.io/probatus/howto/grouped_data.html .

File: probatus/utils/shap_helpers.py [hunk 2]
```diff
@@ -200,3 +201,66 @@ def calculate_shap_importance(shap_values, columns, output_columns_suffix=""):
...
+
+        X_train_cv = X_train.iloc[train_index]
+        y_train_cv = y_train.iloc[train_index]
+
+        model.fit(X_train_cv, y_train_cv)
```
Comment (Matgrb) [inline probatus/utils/shap_helpers.py:244]: Good to check if the model is a search optimization, see ShapRFECV how we dealt with that. 
If it is optimization like RandomizedSearchCV, you have two options:
- Optimize once on the entire dataset, with using CV that was passed by the user. Once you optimize parameters you run each stage of CV with the model with optimized parameters. Thanks to this, at each iteration you end up with model with the same hyperparameters (more stable results over folds), and it is much faster because you don't need to optimize it each time. We chose this option
- Optimize within each CV iteration, which is problematic because you need to do CV inside of CV- which is much slower and less stable over folds.

Handling optimization would be great for this function, because then we can use it inside ShapRFECV and others, otherwise, it will be a standalone function that we don't use anywhere else in probatus.

File: probatus/utils/shap_helpers.py [hunk 3]
```diff
@@ -200,3 +201,66 @@ def calculate_shap_importance(shap_values, columns, output_columns_suffix=""):
...
+
+def get_shap_cv_values(model, X_train, y_train, cv, approximate=False, check_additivity=True, **shap_kwargs):
+    """
+    Calculates SHAP values of a model via cross validation.
+
+    Args:
+        model (binary model):
+            Trained model.
+
+        X_train (pd.DataFrame or np.ndarray):
```
Comment (Matgrb) [inline probatus/utils/shap_helpers.py:214]: I would just call it X and y. Then the train folds fan be called X_train and test folds X_test. 
That is how we called it in other modules e.g. ShapRFECV. It makes sense,  because you can run this CV on any dataset, X_train or all the data you have, just to get an understanding of the dataset.

File: probatus/utils/shap_helpers.py [hunk 4]
```diff
@@ -200,3 +201,66 @@ def calculate_shap_importance(shap_values, columns, output_columns_suffix=""):
...
+        )
+
+        if not shap_values_cv:
+            shap_values_cv = shap_values_cv_split[0]
+            shap_values_cv = [shap_values_cv, shap_values_cv_split[1]]
+        else:
+            shap_values_cv[0] = np.vstack((shap_values_cv[0], shap_values_cv_split[0]))
+            shap_values_cv[1] = np.vstack((shap_values_cv[1], shap_values_cv_split[1]))
+
+    return shap_values_cv
```
Comment (Matgrb) [inline probatus/utils/shap_helpers.py:266]: What is the format in which we return the results now?
What do you think about making it a DataFrame with column and row names? That would be nice, since in all other modules we return dataframes for user to analyse

**Review by Matgrb — CHANGES_REQUESTED** [review]
Overall looks good.

I left some comments, and i have one main worry here:
In order for this function to be usable in other modules it needs to handle hyperparam optimization passed as a model, and be consistent with all applications of this method within probatus e.g. ShapRFECV.fit and possibly Resemblance model in the future. Right now we are adding code that does something similar to what is already there, packaged as a function. I would recommend aligning this function with ShapRFECV.fit, such that we can use this function there. What do you think?

**Comment by Matgrb** [comment]
@rchaves33 Will you be picking this one up? Or shall we close it for now?

There is another PR that will modify some of the functions you are working on. I want to prevent future conflicts ;)

**Comment by ReinierKoops** [comment]
This PR is closed due to inactivity. Please reach out if you want to continue with this PR.
